src/modules/flow/selector/selector.c

```diff
@@ -0,0 +1,688 @@
+ */
+
+#include "sol-flow/selector.h"
+#include "sol-flow-internal.h"
+
+#include "sol-util.h"
+#include "sol-buffer.h"
+
+#include <errno.h>
+
+#define MIN_SIZE(a, b) (((a) < (b)) ? (a) : (b))
+#define MAX_SIZE(a, b) (((a) > (b)) ? (a) : (b))
+
+#define DITCH_NL (true)
+#define KEEP_NL (false)
+
+#define DO_FORMAT (false)
+#define CALC_ONLY (true)
+
+static const char CR = '\r', NL = '\n', SPC = ' ', NUL = '\0';
+static const char *TITLE_TAG = "{title}", *VALUE_TAG = "{value}";
+
+struct selector_data {
+    struct sol_flow_node *node;
+    char *title, *sel_mark, *cursor_mark, *text_mem;
+    char *format, *title_tag, *value_tag;
+    struct sol_buffer text_grid;
+    struct sol_ptr_vector items;
+    uint16_t selection, cursor, n_values;
+    size_t columns, rows;
+    bool circular : 1;
+    bool enabled : 1;
+    bool n_values_done : 1;
+};
+
+static inline size_t
+get_buf_size(struct selector_data *mdata)
+{
+    /* +1 on cols for '\n' and (final) NUL chars */
+    return mdata->rows * (mdata->columns + 1);
+}
+
+static inline void
+buffer_re_init(struct selector_data *mdata)
+{
+    /* We choose to do the ending nul byte ourselves, since we're
+     * doing fixed capacity, to able to set the second to last byte on
+     * the buffer without reallocation attempts. */
+    sol_buffer_init_flags(&mdata->text_grid, mdata->text_mem,
+        get_buf_size(mdata), SOL_BUFFER_FLAGS_FIXED_CAPACITY |
+        SOL_BUFFER_FLAGS_NO_NUL_BYTE);
+    memset(mdata->text_mem, 0, sizeof(*mdata->text_mem));
+}
+
+static int
+selector_open(struct sol_flow_node *node,
+    void *data,
+    const struct sol_flow_node_options *options)
+{
+    int r;
+    struct selector_data *mdata = data;
+    const struct sol_flow_node_type_selector_options *opts =
+        (const struct sol_flow_node_type_selector_options *)options;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(options,
+        SOL_FLOW_NODE_TYPE_SELECTOR_OPTIONS_API_VERSION, -EINVAL);
+
+    mdata->circular = opts->circular;
+    mdata->node = node;
+
+    if (opts->columns.val <= 0) {
+        SOL_WRN("Selector columns number must be a positive integer, "
+            "but %d was given.", opts->columns.val);
+        return -EINVAL;
+    } else
+        mdata->columns = opts->columns.val;
+
+    if (opts->rows.val <= 0) {
+        SOL_WRN("Selector rows number must be a positive integer, "
+            "but %d was given.", opts->rows.val);
+        return -EINVAL;
+    } else
+        mdata->rows = opts->rows.val;
+
+    mdata->format = strdup(opts->format);
+    SOL_NULL_CHECK(mdata->format, -ENOMEM);
+
+    mdata->title_tag = strstr(mdata->format, TITLE_TAG);
+    mdata->value_tag = strstr(mdata->format, VALUE_TAG);
+
+    if (!mdata->value_tag) {
+        SOL_WRN("Bad format, no {value} tag: %s", mdata->format);
+        r = -EINVAL;
+        goto err_format;
+    }
+
+    if (mdata->title_tag > mdata->value_tag) {
+        SOL_WRN("Bad format, {title} tag placet after {value} tag: %s",
```

Comment:
It doesn't seem to be explained on "format" description.
Also, instead of just aborting here I would suggest to use default format.